readthedocs/projects/tasks.py

```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
             self.config = None
 
             setup_successful = self.run_setup(record=record)
-            if setup_successful:
-                self.run_build(record=record, docker=docker)
+            if not setup_successful:
```

Comment:
I _think_ there is something missing here. If the build fails because of the setup at this point: https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R198

we are not raising any exception, so `run_setup` method just returns `False` and the task return `False` but nobody is updating the `self.setup_env.build['error']` as we do when unhandled exception is catched at https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R137 or https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R154

Also, I think that's better to move this line https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R207 inside this if, so we try to update the build state in not so many places

Reply:
Links above have disconnected here, so hard to follow what's happening, but i think i understand this feedback.

So update of the build now always happens on build env `__exit__`, so we can just return false from the task.

In the case that another exception is raised outside the env, we still run `update_build` from the exception.

Reply:
Is this logic ^^ documented somewhere?

Reply:
Yup, on the environment class. I can expand more.